indexer/categories_index.cpp

```diff
@@ -0,0 +1,98 @@
+#include "categories_index.hpp"
+
+#include "base/assert.hpp"
+#include "base/stl_add.hpp"
+#include "base/stl_helpers.hpp"
+#include "base/string_utils.hpp"
+
+#include "std/algorithm.hpp"
+#include "std/set.hpp"
+
+namespace
+{
+void AddAllNonemptySubstrings(my::MemTrie<string, uint32_t> & trie, string const & s,
+                              uint32_t value)
+{
+  ASSERT(!s.empty(), ());
+  for (size_t i = 0; i < s.length(); ++i)
+  {
+    string t;
+    for (size_t j = i; j < s.length(); ++j)
+    {
+      t.push_back(s[j]);
+      trie.Add(t, value);
+    }
+  }
+}
+
+void TokenizeAndAddAllSubstrings(my::MemTrie<string, uint32_t> & trie, string const & s,
+                                 uint32_t value)
+{
+  vector<string> tokens;
+  strings::Tokenize(s, " \n\r\t", MakeBackInsertFunctor(tokens));
```

Comment:
Sorry for the late comment, but may be it's worth to use NormalizeAndSimplifyString + search tokenizer here? Up to you.